server/model/splittable/split_table_recognizer.py

```diff
@@ -0,0 +1,132 @@
+import pickle
+
+from server.service.data.data import *
+
+
+class SplitTableRecognizer:
+    def process(self, document: Document) -> Document:
+        pass
+
+    def merge(self, table1: Table, table2: Table) -> Table:
+        for row in table2.rows:
+            table1.add_row(row)
+
+        return table1
+
+    def column_number_diff(self, table1, table2):
+        cells1 = table1.rows[0].cells
+        cells2 = table2.rows[0].cells
+        return len(cells2) - len(cells1)
+
+
+class SplitTableModel(SplitTableRecognizer):
+
+    def process(self, document: Document) -> Document:
+
+        for i in range(len(document.tables) - 1):
+            table1 = document.tables[i]
+            table2 = document.tables[i + 1]
+
+            if table2.page_index - table1.page_index != 1:
+                continue
+
+            if self.column_number_diff(table1, table2) != 0:
+                continue
+
+            X = self.calculate_x(document, table1, table2)
+
+            if self.predict(X):
+                document.tables[i + 1] = self.merge(table1, table2)
+                document.remove_table(i)
+
+        return document
+
+    def calculate_x(self, document, table1, table2):
+        return [[
+            (document.height - table1.bbox.lower_right.y) / table1.bbox.height(),
+            table2.bbox.upper_left.y / table2.bbox.height(),
+            abs(table1.bbox.width() - table2.bbox.width()) / table1.bbox.width(),
+            abs(table1.bbox.upper_left.x - table2.bbox.upper_left.x) / table1.bbox.width(),
+            abs(table1.bbox.lower_right.x - table2.bbox.lower_right.x) / table1.bbox.width()
+        ]]
+
+    def predict(self, X: list) -> bool:
+        model = pickle.load(open("model.sav", 'rb'))
+        result = model.predict(X)
+        return result[0]
+
+
+class SplitTableHeuristic(SplitTableRecognizer):
+
+    def process(self, document: Document) -> Document:
+
+        for i in range(len(document.tables) - 1):
+            table1 = document.tables[i]
+            table2 = document.tables[i + 1]
+            if self.check_merge(table1, table2, document.height):
+                document.tables[i + 1] = self.merge(table1, table2)
+                document.remove_table(i)
+
+        return document
+
+    def check_merge(self, table1: Table, table2: Table, page_height: int) -> bool:
+
+        eps = 0.05 * table1.bbox.width()
+        margin = 0.08 * table1.bbox.height()
```

Comment:
Czy te wartości dobrałaś empirycznie na jakimś datasecie, żeby były najlepsze wyniki? Czy to bardziej na oko?

Reply:
Na oko na razie, zastanawiałam się czy nie zostawić tego w ogóle użytkownikowi do wybrania, ale i tak muszę się zastanowić nad domyślnymi jeszcze - sprawdzę jak to wygląda na tych podzielonych stronach (ile to %)

Reply:
Dobry pomysł, żeby dodać to do konfiguracji :)